### PR TITLE
Add tests for druid and c3p0

### LINKED ISSUE #27988: In c3p0, it is necessary to call the get method to obtain the value through reflection, not through the reflection attribute
### Which version of ShardingSphere did you use?
-- 5.4.1-SNAPSHOT

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
-- ShardingSphere-JDBC

### Expected behavior
-- Pass all single tests

### Actual behavior
-- actual.getProperties() cannot pass, no corresponding value is set
### Reason analyze (If you can)
in c3p0, class for AbstractComboPooledDataSource, The method of setting properties is as follows
```
public Properties getProperties() {
        return this.dmds.getProperties();
    }

    public void setProperties(Properties properties) {
        if (diff(this.dmds.getProperties(), properties)) {
            this.dmds.setProperties(properties);
            this.resetPoolManager(false);
        }

    }
```
If we need to get and set the value of properties through reflection, we need to call these two methods through reflection instead of directly setting the properties attribute.

### Example codes for reproduce this issue (such as a github link).
```

class C3P0DataSourcePoolCreatorTest {
    @Test
    void assertCreateDataSource() {
        ComboPooledDataSource actual = (ComboPooledDataSource) DataSourcePoolCreator.create(new DataSourceProperties(ComboPooledDataSource.class.getName(), createDataSourceProperties()));
        assertThat(actual.getJdbcUrl(), is("jdbc:mock://127.0.0.1/foo_ds"));
         assertThat(actual.getUser(), is("root"));
         assertThat(actual.getPassword(), is("root"));
         // c3p0 will automatically add username and password to properties  {@see DriverManagerDataSource#setUser}
        assertThat(actual.getProperties(), is(
                PropertiesBuilder.build(new PropertiesBuilder.Property("foo", "foo_value"), new PropertiesBuilder.Property("bar", "bar_value"),
                        new PropertiesBuilder.Property("password", "root"), new PropertiesBuilder.Property("user", "root")
                        )));
    }

    private Map<String, Object> createDataSourceProperties() {
        Map<String, Object> result = new HashMap<>();
        result.put("url", "jdbc:mock://127.0.0.1/foo_ds");
        result.put("driverClassName", MockedDataSource.class.getName());
        result.put("username", "root");
        result.put("password", "root");
        result.put("properties", PropertiesBuilder.build(new PropertiesBuilder.Property("foo", "foo_value"),new PropertiesBuilder.Property("bar", "bar_value")));
        return result;
    }
```

### PR DESCRIPTION
**Pull request by daleyzou**
Add tests for druid and c3p0l (#27894)
Fixes #27988
* test for Support more data source connection pool

* Fix setting properties using reflection

### CODE REVIEW

File: infra/datasource/type/c3p0/src/test/java/org/apache/shardingsphere/infra/datasource/c3p0/creator/C3P0DataSourcePoolCreatorTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+class C3P0DataSourcePoolCreatorTest {
+    @Test
```
Comment (Pace2Car) [inline infra/datasource/type/c3p0/src/test/java/org/apache/shardingsphere/infra/datasource/c3p0/creator/C3P0DataSourcePoolCreatorTest.java:34]: need a blank line.

File: infra/datasource/type/c3p0/src/test/java/org/apache/shardingsphere/infra/datasource/c3p0/creator/C3P0DataSourcePoolCreatorTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    void assertCreateDataSource() {
+        ComboPooledDataSource actual = (ComboPooledDataSource) DataSourcePoolCreator.create(new DataSourceProperties(ComboPooledDataSource.class.getName(), createDataSourceProperties()));
+        assertThat(actual.getJdbcUrl(), is("jdbc:mock://127.0.0.1/foo_ds"));
+        assertThat(actual.getUser(), is("root"));
+        assertThat(actual.getPassword(), is("root"));
+        // c3p0 will automatically add username and password to properties  {@see DriverManagerDataSource#setUser}
```
Comment (Pace2Car) [inline infra/datasource/type/c3p0/src/test/java/org/apache/shardingsphere/infra/datasource/c3p0/creator/C3P0DataSourcePoolCreatorTest.java:40]: Do not leave comments here.

File: infra/util/src/main/java/org/apache/shardingsphere/infra/util/reflection/ReflectionUtils.java [hunk 3]
```diff
@@ -134,4 +137,36 @@ public static <T> T invokeMethod(final Method method, final Object target, final
...
+        }
+        
+        return Optional.empty();
+    }
+    
+    private static Method findMethod(final Class<?> clazz, final String methodName, final Class<?>... parameterTypes) {
+        try {
+            return clazz.getMethod(methodName, parameterTypes);
+        } catch (NoSuchMethodException e) {
+            // Try superclass if method not found in this class
```
Comment (Pace2Car) [inline infra/util/src/main/java/org/apache/shardingsphere/infra/util/reflection/ReflectionUtils.java:164]: Do not leave comments.

File: infra/datasource/type/c3p0/src/test/java/org/apache/shardingsphere/infra/datasource/c3p0/creator/C3P0DataSourcePoolCreatorTest.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+class C3P0DataSourcePoolCreatorTest {
+    
+    /**
```
Comment (Pace2Car) [inline infra/datasource/type/c3p0/src/test/java/org/apache/shardingsphere/infra/datasource/c3p0/creator/C3P0DataSourcePoolCreatorTest.java:35]: no need comments for test methods.

File: infra/util/src/main/java/org/apache/shardingsphere/infra/util/reflection/ReflectionUtils.java [hunk 5]
```diff
@@ -134,4 +137,36 @@ public static <T> T invokeMethod(final Method method, final Object target, final
...
+     * @return field value
+     */
+    public static <T> Optional<T> getFieldValueByGetMethod(final Object target, final String fieldName) {
+        String getterName = GETTER_PREFIX + CaseFormat.LOWER_CAMEL.to(CaseFormat.UPPER_CAMEL, fieldName);
+        final Method method = findMethod(target.getClass(), getterName);
+        if (method != null) {
+            T value = invokeMethod(method, target);
+            return Optional.ofNullable(value);
+        }
+
```
Comment (Pace2Car) [inline infra/util/src/main/java/org/apache/shardingsphere/infra/util/reflection/ReflectionUtils.java:156]: redundant blank line.

File: infra/datasource/core/src/main/java/org/apache/shardingsphere/infra/datasource/pool/creator/DataSourceReflection.java [hunk 6]
```diff
@@ -170,13 +174,15 @@ public void addDefaultDataSourceProperties() {
         }
         ConnectionProperties connectionProps = DatabaseTypedSPILoader.getService(ConnectionPropertiesParser.class, DatabaseTypeFactory.get(jdbcUrl.get())).parse(jdbcUrl.get(), null, null);
         Properties queryProps = connectionProps.getQueryProperties();
+        Properties properties = jdbcConnectionProps.get();
```
Comment (Pace2Car) [inline infra/datasource/core/src/main/java/org/apache/shardingsphere/infra/datasource/pool/creator/DataSourceReflection.java:177]: Variable name using `jdbcProps` would be better.

File: infra/datasource/core/src/main/java/org/apache/shardingsphere/infra/datasource/pool/creator/DataSourceReflection.java [hunk 7]
```diff
@@ -158,8 +161,9 @@ private Optional<Method> findSetterMethod(final String fieldName) {
     
     /**
      * Add default data source properties.
+     * @param dataSourcePoolMetaData data source pool meta data
```
Comment (Pace2Car) [inline infra/datasource/core/src/main/java/org/apache/shardingsphere/infra/datasource/pool/creator/DataSourceReflection.java:164]: A blank line is required before.

File: infra/datasource/core/src/main/java/org/apache/shardingsphere/infra/datasource/pool/metadata/DataSourcePoolMetaDataReflection.java [hunk 8]
```diff
@@ -36,19 +36,31 @@ public final class DataSourcePoolMetaDataReflection {
     
     /**
      * Get JDBC URL.
+     * Get it directly through the attribute first, and get it through the get method when it is empty
```
Comment (Pace2Car) [inline infra/datasource/core/src/main/java/org/apache/shardingsphere/infra/datasource/pool/metadata/DataSourcePoolMetaDataReflection.java:39]: unnecessary comments.

File: infra/datasource/core/src/main/java/org/apache/shardingsphere/infra/datasource/pool/metadata/DataSourcePoolMetaDataReflection.java [hunk 9]
```diff
@@ -36,19 +36,31 @@ public final class DataSourcePoolMetaDataReflection {
...
+            return jdbcUrl;
+        } else {
+            return ReflectionUtils.getFieldValueByGetMethod(targetDataSource, dataSourcePoolFieldMetaData.getJdbcUrlFieldName());
+        }
     }
     
     /**
      * Get JDBC connection properties.
-     * 
+     * Get it directly through the attribute first, and get it through the get method when it is empty
```
Comment (Pace2Car) [inline infra/datasource/core/src/main/java/org/apache/shardingsphere/infra/datasource/pool/metadata/DataSourcePoolMetaDataReflection.java:54]: unnecessary comments.

**Review by RaigorJiang — APPROVED** [review]

**Comment by RaigorJiang** [comment]
@daleyzou @Pace2Car  merged, thank you!

**Pull request merged**